lib/Parse/ParseDecl.cpp

```diff
@@ -5743,6 +5740,41 @@ setOriginalDeclarationForDifferentiableAttributes(DeclAttributes attrs,
     const_cast<DerivativeAttr *>(attr)->setOriginalDeclaration(D);
 }
 
+/// Determine the declaration parsing options to use when parsing the decl in
+/// the given context.
+static Parser::ParseDeclOptions getParseDeclOptions(DeclContext *DC) {
+  using ParseDeclOptions = Parser::ParseDeclOptions;
+
+  if (DC->isModuleScopeContext())
+    return Parser::PD_AllowTopLevel;
+
+  if (DC->isLocalContext())
+    return Parser::PD_Default;
+
+  auto decl = DC->getAsDecl();
```

Comment:
Would it be worth checking if `decl != nullptr`? I don’t know when the parser might have a `DeclContext` that’s not a `Decl` (e.g. a closure) but seems like a potential source of crashes.

Reply:
This _should_ always be non-null. All non-decl `DeclContext` cases should be handled above (i.e. local context, or mode scope context)